### PR TITLE
#2935 

### PR DESCRIPTION
**Pull request by zzp198-old**
### What is the new feature?

When player saving fails, plr files and backups will not be modified and the game will be kicked out.

### Why should this be part of tModLoader?

This can ensure that the player's data will not be lost when the player has enabled the problematic mod.

### Are there alternative designs?

No.

### Sample usage for the new feature

[SaveCrash.zip](https://github.com/tModLoader/tModLoader/files/9596657/SaveCrash.zip)
This is a problematic mod source code. It will modify the first item of the inventory every 5 seconds and save it. Every time save it, there is a 1/5 probability of an error. Enable it until an error is reported, and then disable it. Check the player's first item to verify.

### CODE REVIEW

**Comment by zzp198** [comment]
Sorry, I changed a lot when formatting the code.

File: patches/tModLoader/Terraria/Main.TML.cs [hunk 1]
```diff
@@ -195,7 +195,7 @@ public partial class Main
 
 			base.Content = new TMLContentManager(Content.ServiceProvider, vanillaContentFolder, AlternateContentManager);
 		}
-		
+
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Main.TML.cs:198]: Please check your patches and try and avoid adjusting formatting in files unrelated to your PR

File: patches/tModLoader/Terraria/Main.cs.patch [hunk 2]
```diff
@@ -470,7 +470,7 @@
  			}
  		}
  
-+		/* Disabled, tML only supports HiDef		
++		/* Disabled, tML only supports HiDef
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Main.cs.patch:473]: as above, this change doesn't belong in this PR

File: patches/tModLoader/Terraria/Player.cs.patch [hunk 3]
```diff
@@ -5682,19 +5682,52 @@
...
+ 			catch (Exception exception) {
+ 				FancyErrorPrinter.ShowFileSavingFailError(exception, playerFile.Path);
++				// If there is a problem in saving, (kick out the player and) show message
++				var errorBuilder = new StringBuilder();
++				errorBuilder.AppendLine(Language.GetTextValue("An error occurred when saving the player.")); // TODO use key
++				errorBuilder.AppendLine(Language.GetTextValue("The original Error below:")); // TODO use key
++				errorBuilder.Append(exception);
++
++				var flag = false;
++				if (!flag && !Main.gameMenu) {
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Player.cs.patch:5692]: Please call `WorldGen.SaveAndQuit` instead.

File: patches/tModLoader/Terraria/Player.cs.patch [hunk 4]
```diff
@@ -5682,19 +5682,52 @@
...
++					// Single & Multiplayer will quit. Server continues to run.
++					if (Main.netMode != NetmodeID.Server) {
++						WorldGen.JustQuit();
++						if (Main.netMode == NetmodeID.SinglePlayer) {
++							Main.menuMode = MenuID.Title;
++						}
++					}
++				}
++
++				Utils.ShowFancyErrorMessage(errorBuilder.ToString(), Main.menuMode); // Return to the previous menu.
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Player.cs.patch]: Suggest:
`Utils.ShowFancyErrorMessage($"{Language.GetTextValue("tModLoader.PlayerSaveFail")}\n\n{e}", Main.menuMode)`

`"PlayerLoadWorldFail": "An error occurred while saving the player"`

File: patches/tModLoader/Terraria/Player.cs.patch (hunk 3 above)
Comment (zzp198-old) [reply patches/tModLoader/Terraria/Player.cs.patch:5692]: WorldGen.SaveAndQuit will call Player.SavePlayer, cause an endless loop
